clippy_lints/src/unnecessary_self_imports.rs

```diff
@@ -0,0 +1,177 @@
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::def_id::DefId;
+use rustc_span::symbol::Ident;
+use rustc_span::{BytePos, Span};
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for imports ending in `::{self};`.
+    ///
+    /// **Why is this bad?** In most cases, this can be written much more cleanly by omitting `self`.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// use std::io::{self};
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::io;
+    /// ```
+    pub UNNECESSARY_SELF_IMPORTS,
+    style,
+    "imports ending in `self`, which can be omitted"
+}
+
+declare_lint_pass!(UnnecessarySelfImports => [UNNECESSARY_SELF_IMPORTS]);
+
+impl<'tcx> LateLintPass<'tcx> for UnnecessarySelfImports {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'_>) {
+        if_chain! {
+            if !in_macro(item.span);
+            if let ItemKind::Use(path, UseKind::ListStem) = &item.kind;
+            if let Some(ident) = is_self_import(cx, item, path);
+            if let Some(Res::Def(DefKind::Mod, def_id)) = path.segments[path.segments.len() - 2].res;
+            if let Some(last) = path.segments.last();
+            if !mod_contains_item(cx, def_id, last.ident) || !item_is_in_scope(cx, ident, path.span);
+            then {
+                let mut applicability = Applicability::MachineApplicable;
+                let snippet = if ident == path.segments[path.segments.len() - 1].ident {
+                    let adjusted_span = path.span.with_hi(path.span.hi() - BytePos(8));
+                    format!(
+                        "{}",
+                        snippet_with_applicability(cx, adjusted_span, "..", &mut applicability)
+                    )
+                } else {
+                    let adjusted_span = path.span.until(ident.span);
+                    let adjusted_span = adjusted_span.with_hi(adjusted_span.hi() - BytePos(11));
+                    format!(
+                        "{} as {}",
+                        snippet_with_applicability(cx, adjusted_span, "..", &mut applicability),
+                        snippet_with_applicability(cx, ident.span, "..", &mut applicability)
+                    )
+                };
+                span_lint_and_sugg(
+                    cx,
+                    UNNECESSARY_SELF_IMPORTS,
+                    path.span,
+                    "import ending with `self`",
+                    "consider omitting `::{self}`",
+                    snippet,
+                    applicability,
+                );
+            }
+        }
+    }
+}
+
+fn is_self_import(cx: &LateContext<'_>, current_import: &Item<'_>, current_import_path: &Path<'_>) -> Option<Ident> {
+    let definitions = cx.tcx.hir().definitions();
+    let mut amt = 0;
+    let mut ident: Option<Ident> = None;
+
+    for def in definitions.iter_local_def_id() {
```

Comment:
I don't really get the point of this function. Why do you have to check **all** definitions in a `crate`? Couldn't you just check if the last path segment is `self`?

Reply:
I could be missing something, but when `self` is included in an import it's not included as a path segment, so the last path segment is whatever `self` refers to. My thinking was to get the last path segment from a `ListStem` item and compare that against the last path segment in `Single` items contained in that scope. If they're the same, then the import includes `self`. I'll look into ways to optimize it though, because I definitely don't need all the defnitions in the crate.

Reply:
IIRC I did something similar in the `wildcard_imports` lint to build the suggestion. You shouldn't have to go over every definition. Let me look that up.